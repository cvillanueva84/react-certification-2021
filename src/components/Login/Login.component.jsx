import React, { useState, useRef } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import Alert from '@material-ui/lab/Alert';
import { useAuth } from '../../providers/Auth';
import loginApi from './Login.api';
// import { useHistory } from 'react-router-dom';

const Login = ({ show, toggleShow }) => {
  const [error, setError] = useState(false);
  const [msgError, setMsgError] = useState('');
  const [disabled, setDisabled] = useState(false);

  const userName = useRef();
  const passWord = useRef();
  // const history = useHistory();

  const { login } = useAuth();

  // console.log('Login auth', authenticated);

  const handleReset = () => {
    toggleShow();
  };

  const handleSubmit = async (event) => {
    event.preventDefault();

    setDisabled(true);
    if (error) setError(false);

    const user = userName.current.value;
    const pass = passWord.current.value;

    try {
      const solveUser = await loginApi(user, pass);
      console.log(solveUser);
      login();
    } catch (err) {
      setMsgError(err.message);
      setError(true);
    }
    setDisabled(false);
  };

  return (
    <Dialog open={show} onClose={handleReset} aria-labelledby="form-dialog-title">
      <DialogTitle id="form-dialog-title">Login</DialogTitle>
      <DialogContent>
        {error ? <Alert severity="error">{msgError}</Alert> : null}
        <form onSubmit={handleSubmit} onReset={handleReset}>
          <TextField
            autoFocus
            margin="dense"
            label="Username"
            type="text"
            inputRef={userName}
            fullWidth
            disabled={disabled}
          />
          <TextField
            margin="dense"
            label="Password"
            type="password"
            inputRef={passWord}
            fullWidth
            disabled={disabled}
          />
          <DialogActions>
            <Button type="reset" color="primary">
              Cancel
            </Button>
            <Button type="submit" color="primary">
              Login
            </Button>
          </DialogActions>
        </form>
      </DialogContent>
    </Dialog>
  );
};

export default Login;
