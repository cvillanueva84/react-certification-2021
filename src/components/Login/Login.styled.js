import styled from 'styled-components';

export const Container = styled.div`
  background-color: ${(props) => props.theme.body};
  padding-top: 48px;
  min-height: calc(100vh - 64px);
  display: flex;
  justify-content: center;
`;

export const CardLogin = styled.div`
  background-color: ${(props) => props.theme.cardBg};
  color: ${(props) => props.theme.color};
  display: flex;
  flex-direction: column;
  align-items: center;
  border: 0;
  padding: 24px;
  outline: 0;
  width: 445px;
  height: 330px;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0, 0, 0, 0.2), 0px 1px 1px 0px rgba(0, 0, 0, 0.14),
    0px 1px 3px 0px rgba(0, 0, 0, 0.12);
`;

export const CardTitle = styled.h2`
  text-align: center;
  padding-bottom: 16px;
`;

export const TextLabel = styled.label`
  color: ${(props) => props.theme.color};
  font-weight: 400;
`;

export const TextField = styled.input`
  outline: 0;
  width: 100%;
  padding: 12px 20px;
  margin: 8px 0;
  display: inline-block;
  border: 1px solid #ccc;
  border-radius: 4px;
  box-sizing: border-box;
  color: inherit;
  background: transparent;
  font-weight: 400;

  &:focus {
    border: 2px solid ${(props) => props.theme.btn};
  }
  &:hover {
    border: 1px solid ${(props) => props.theme.btn};
  }
`;

export const Button = styled.button`
  background-color: ${(props) => props.theme.btn};
  border: none;
  color: white;
  padding: 15px 32px;
  text-align: center;
  text-decoration: none;
  display: inline-block;
  font-size: 16px;
  margin-top: 16px;
  cursor: pointer;
  border-radius: 4px;
  width: 100%;
`;
