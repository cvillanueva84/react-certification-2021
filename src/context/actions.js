import Types from '../types';

export const contextValueActions = (dispatch) => ({
  search: (payload) => dispatch({ type: Types.SEARCH, payload }),
  theme: () => dispatch({ type: Types.THEME }),
  auth: (payload) => dispatch({ type: Types.AUTH, payload }),
  authLogout: () => dispatch({ type: Types.LOGOUT }),
  videoList: (payload) => dispatch({ type: Types.VIDEO_LIST, payload }),
});
