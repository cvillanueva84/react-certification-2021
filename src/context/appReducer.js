import { types } from '../types/types';

export const appReducer = (state = {}, action) => {
  switch (action.type) {
    case types.search:
      return {
        ...state,
        ...action.payload,
      };
    case types.showVideoDetails:
      return {
        ...state,
        showVideoDetails: true,
      };
    case types.showVideoList:
      return {
        ...state,
        showVideoDetails: false,
      };
    default:
      return state;
  }
};
