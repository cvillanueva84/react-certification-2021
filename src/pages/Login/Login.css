h2 {
  font-size: 1.75rem;
}

input {
  background-image: none;
  border: none;
  font: inherit;
  margin: 0;
  padding: 0;
  transition: all 0.3s;
}

svg {
  height: auto;
  max-width: 100%;
  vertical-align: middle;
}

/* ---------- ALIGN ---------- */
.align {
  height: 85vh;
  display: flex;
  align-items: center;
  justify-content: center;
}

/* ---------- BUTTON ---------- */

.button {
  background-color: #33cc77;
  margin: 0 5px 0 5px;
  color: #fff;
  padding: 0.25em 1.5em;
  cursor: pointer;
}

.btn-cancel {
  background-color: #bfc5c2;
}

.button:focus,
.button:hover {
  background-color: #28ad63;
}

.btn-cancel:focus,
.btn-cancel:hover {
  background-color: #8b8f8d;
}

/* ---------- ICONS ---------- */
.icons {
  display: none;
}

.icon {
  fill: currentcolor;
  display: inline-block;
  height: 1em;
  width: 1em;
}

/* ---------- LOGIN ---------- */

.file {
  display: flex;
}

.login {
  width: 400px;
}

.login__header {
  background-color: #f95252;
  border-top-left-radius: 1.25em;
  border-top-right-radius: 1.25em;
  color: #fff;
  padding: 1.25em 1.625em;
}

.login__header :first-child {
  margin-top: 0;
}

.login__header :last-child {
  margin-bottom: 0;
}

.login h2 .icon {
  margin-right: 14px;
}

.login__form {
  background-color: #fff;
  border-bottom-left-radius: 1.25em;
  border-bottom-right-radius: 1.25em;
  color: #777;
  display: grid;
  gap: 0.875em;
  padding: 1.25em 1.625em;
}

.login input {
  border-radius: 0.1875em;
}

.login p {
  color: black;
  padding: 5px;
  background-color: rgb(248, 225, 225);
  border-radius: 5px;
  text-align: center;
  margin: 0;
}

.login input[type='text'],
.login input[type='password'] {
  background-color: #eee;
  color: #777;
  padding: 0.25em 0.625em;
  width: 100%;
}

.login input[type='submit'],
.login input[type='button'] {
  display: block;
  margin: 0 auto;
}
