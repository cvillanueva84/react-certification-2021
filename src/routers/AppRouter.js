import React, { useContext } from 'react';
import { BrowserRouter as Router, Switch } from "react-router-dom";
import { GlobalContext } from '../components/Context';
import { LoginView } from '../components/LoginView';
import { DashboardRoutes } from './DashboardRoutes';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

export const AppRouter = () => {

    const { myStateReducer } = useContext(GlobalContext);

    return (
        <Router>
            <>
                <Switch>
                    <PublicRoute 
                        exact path="/login" 
                        component={LoginView} 
                        isAuthenticated={myStateReducer.logged}
                    />

                    <PrivateRoute 
                        path="/" 
                        component={DashboardRoutes} 
                        isAuthenticated={myStateReducer.logged}
                    />
                </Switch>
            </>
        </Router>
    )
}
