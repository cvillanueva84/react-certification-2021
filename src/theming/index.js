const lightTheme = {
  body: '#E5E5E5',
  cardBg: '#fff',
  cardBgHover: 'rgb(240, 240, 240)',
  color: '#000',
  headerBg: '#1C5476',
  textTiny: 'rgb(70, 70, 70)',
};

const darkTheme = {
  body: '#1e1e2f',
  cardBg: '#27293d',
  cardBgHover: '#7b1fa2',
  color: '#fff',
  headerBg: '#7b1fa2',
  textTiny: '#fff',
};

export { lightTheme, darkTheme };
