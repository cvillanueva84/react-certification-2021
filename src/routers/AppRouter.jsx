import React, { useContext, useEffect, useState } from 'react';
import { Switch, Redirect, Route } from 'react-router-dom';

import Favorites from '../components/Favorites';
import FavoriteVideoDetails from '../components/FavoriteVideoDetails';
import Home from '../components/Home/Home.component';
import Login from '../components/Login/Login.component';
import VideoDetails from '../components/VideoDetails';
import { AppContext } from '../context/appContext';
import { firebase } from '../firebase/firebase.config';
import { types } from '../types/types';
import { PrivateRoute } from './PrivateRoute';

export const AppRouter = () => {
  const { dispatch } = useContext(AppContext);
  const [isLoggedIn, setIsLoggedIn] = useState(false);

  useEffect(() => {
    firebase.auth().onAuthStateChanged(async (user) => {
      if (user?.uid) {
        dispatch({
          type: types.auth,
          payload: {
            isAuthenticated: true,
            auth: {
              uid: user.uid,
              email: user.email,
            },
          },
        });
        setIsLoggedIn(true);
      } else {
        setIsLoggedIn(false);
      }
    });
  }, [dispatch]);

  return (
    <Switch>
      <Route path="/" exact component={Home} />
      <Route
        path="/login"
        exact
        component={() => (isLoggedIn ? <Redirect to="/" /> : <Login />)}
      />
      <Route path="/video/:id" exact component={VideoDetails} />
      <PrivateRoute
        isAuthenticated={isLoggedIn}
        component={Favorites}
        path="/favorites"
        exact
      />
      <PrivateRoute
        isAuthenticated={isLoggedIn}
        component={FavoriteVideoDetails}
        path="/favorite/:id"
        exact
      />
      <Redirect to="/" />
    </Switch>
  );
};
