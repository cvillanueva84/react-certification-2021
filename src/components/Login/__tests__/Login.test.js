import React from 'react';
import ReactDOM from 'react-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { HashRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import Login from '../index';
import { AppContext } from '../../../context/appContext';
import { darkTheme, lightTheme } from '../../../theming';
import { GlobalStyles } from '../../../GlobalStyles.styled';
import { AppRouter } from '../../../routers/AppRouter';

describe('Testing Login component', () => {
  const initialState = {
    search: 'wizeline',
    theme: true,
    isAuthenticated: false,
    auth: {},
    videoList: [],
  };
  const contextValue = {
    dispatch: jest.fn(),
    state: {
      search: 'wizeline',
      theme: true,
      isAuthenticated: false,
      auth: {},
      videoList: [],
    },
  };
  let tree;

  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((element) => {
      return element;
    });
  });

  afterEach(() => {
    ReactDOM.createPortal.mockClear();
  });

  beforeEach(() => {
    tree = render(
      <AppContext.Provider value={contextValue}>
        <ThemeProvider theme={initialState.theme ? lightTheme : darkTheme}>
          <GlobalStyles />
          <HashRouter>
            <AppRouter />
            <Login />
          </HashRouter>
        </ThemeProvider>
      </AppContext.Provider>
    );
  });

  it('should take a snapshop', () => {
    expect(tree).toMatchSnapshot();
  });

  it('should change input', () => {
    const input = screen.getByPlaceholderText('Email');
    fireEvent.change(input, { target: { value: 'example@gmail.com' } });

    expect(input.value).toBe('example@gmail.com');
  });
});
