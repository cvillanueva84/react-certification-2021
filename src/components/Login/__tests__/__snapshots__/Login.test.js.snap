// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login component should take a snapshop 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-iCoGMd cMlkwg"
      >
        <div>
          <h2
            center="true"
            class="sc-iemWCZ hsyMGv title py-1"
            weight="300"
          >
            Welcome to the Challenge!
          </h2>
          <div
            class="sc-crzoAE ebjTvX"
          >
            <div
              class="sc-dIsUp fFrSBm"
            >
              <div
                class="sc-pNWdM hCqdrl"
              />
              <div
                class="sc-jrsJWt kkAfPJ"
                data-testid="loader"
              >
                <div
                  class="sc-kEqXSa juaPy"
                />
              </div>
              
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-iCoGMd FLRok"
      >
        <div>
          <div
            class="sc-bdnxRM ioJUKE"
          >
            <ul
              class="sc-gtsrHT eluVUp"
            >
              <li
                class="sc-dlnjwi jRVDvi"
              >
                Login
              </li>
              <li
                class="sc-dlnjwi KvkCT"
              >
                Register
              </li>
            </ul>
            <div
              class="sc-hKFxyN gsblKA"
            >
              <form
                novalidate=""
              >
                <h3
                  center="true"
                  class="sc-iemWCZ eaPdlT mt-2 mb-1"
                >
                  Login Form
                </h3>
                
                <label
                  class="sc-jSFjdj SzRKi"
                  for="email"
                >
                  Email:
                  <input
                    class="sc-gKAaRy gJKjrP"
                    name="email"
                    placeholder="Email"
                    type="email"
                  />
                </label>
                <label
                  class="sc-jSFjdj SzRKi"
                  for="password"
                >
                  Password:
                  <input
                    class="sc-gKAaRy gJKjrP"
                    name="password"
                    placeholder="Password"
                    type="password"
                  />
                </label>
                <button
                  class="sc-iqAclL gFdvnG"
                  color="primary"
                  type="submit"
                >
                  Login
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-iCoGMd cMlkwg"
    >
      <div>
        <h2
          center="true"
          class="sc-iemWCZ hsyMGv title py-1"
          weight="300"
        >
          Welcome to the Challenge!
        </h2>
        <div
          class="sc-crzoAE ebjTvX"
        >
          <div
            class="sc-dIsUp fFrSBm"
          >
            <div
              class="sc-pNWdM hCqdrl"
            />
            <div
              class="sc-jrsJWt kkAfPJ"
              data-testid="loader"
            >
              <div
                class="sc-kEqXSa juaPy"
              />
            </div>
            
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-iCoGMd FLRok"
    >
      <div>
        <div
          class="sc-bdnxRM ioJUKE"
        >
          <ul
            class="sc-gtsrHT eluVUp"
          >
            <li
              class="sc-dlnjwi jRVDvi"
            >
              Login
            </li>
            <li
              class="sc-dlnjwi KvkCT"
            >
              Register
            </li>
          </ul>
          <div
            class="sc-hKFxyN gsblKA"
          >
            <form
              novalidate=""
            >
              <h3
                center="true"
                class="sc-iemWCZ eaPdlT mt-2 mb-1"
              >
                Login Form
              </h3>
              
              <label
                class="sc-jSFjdj SzRKi"
                for="email"
              >
                Email:
                <input
                  class="sc-gKAaRy gJKjrP"
                  name="email"
                  placeholder="Email"
                  type="email"
                />
              </label>
              <label
                class="sc-jSFjdj SzRKi"
                for="password"
              >
                Password:
                <input
                  class="sc-gKAaRy gJKjrP"
                  name="password"
                  placeholder="Password"
                  type="password"
                />
              </label>
              <button
                class="sc-iqAclL gFdvnG"
                color="primary"
                type="submit"
              >
                Login
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
